internal/runtimehandlerhooks/high_performance_hooks.go

```diff
@@ -356,6 +417,176 @@ func setCPUQuota(cpuMountPoint, parentDir string, c *oci.Container, enable bool)
+// value so it can be restored later. If the governor is an empty string, the original
+// scaling_governor value is restored.
+func setCPUFreqGovernor(c *oci.Container, governor, cpuDir, cpuSaveDir string) error {
+	lspec := c.Spec().Linux
+	if lspec == nil ||
+		lspec.Resources == nil ||
+		lspec.Resources.CPU == nil ||
+		lspec.Resources.CPU.Cpus == "" {
+		return errors.Errorf("find container %s CPUs", c.ID())
+	}
+
+	cpus, err := cpuset.Parse(lspec.Resources.CPU.Cpus)
+	if err != nil {
+		return err
+	}
+
+	for _, cpu := range cpus.ToSlice() {
+		if governor != "" {
+			// retrieve the current scaling governor
+			cpuScalingGovernorFile := fmt.Sprintf("%s/cpu%d/cpufreq/scaling_governor", cpuDir, cpu)
+			content, err := ioutil.ReadFile(cpuScalingGovernorFile)
+			if err != nil {
+				return err
+			}
+			governorOrig := strings.TrimSpace(string(content))
+
+			// get available cpu scaling governors
+			cpuScalingAvailableGovernorFile := fmt.Sprintf("%s/cpu%d/cpufreq/scaling_available_governors", cpuDir, cpu)
+			content, err = ioutil.ReadFile(cpuScalingAvailableGovernorFile)
+			if err != nil {
+				return err
+			}
+			availableGovernors := strings.Split(strings.TrimSpace(string(content)), " ")
+
+			// is the scaling governor supported?
+			governorFound := false
+			for _, availableGovernor := range availableGovernors {
+				if availableGovernor == governor {
+					governorFound = true
+					break
+				}
+			}
+			if !governorFound {
+				return errors.Errorf("governor %s not available for cpu %d", governor, cpu)
+			}
+
+			// save the current governor so we can restore it later
+			cpuFreqSaveDir := fmt.Sprintf("%s/cpu%d/cpufreq", cpuSaveDir, cpu)
+			err = os.MkdirAll(cpuFreqSaveDir, 0o750)
+			if err != nil {
+				return err
+			}
+			cpuScalingGovernorFileOrig := fmt.Sprintf("%s/scaling_governor", cpuFreqSaveDir)
+			err = ioutil.WriteFile(cpuScalingGovernorFileOrig, []byte(governorOrig), 0o644)
+			if err != nil {
+				return err
+			}
+
+			// update the governor
+			err = ioutil.WriteFile(cpuScalingGovernorFile, []byte(governor), 0o644)
+			if err != nil {
+				return err
+			}
+		} else {
+			// retrieve the original scaling governor
+			cpuScalingGovernorFileOrig := fmt.Sprintf("%s/cpu%d/cpufreq/scaling_governor", cpuSaveDir, cpu)
+			content, err := ioutil.ReadFile(cpuScalingGovernorFileOrig)
+			if err != nil {
+				if errors.Is(err, os.ErrNotExist) {
+					// the governor may have already been restored by a previous invocation of the hook
+					return nil
+				}
+				return err
+			}
+			governorOrig := strings.TrimSpace(string(content))
+
+			// restore the original governor
+			cpuScalingGovernorFile := fmt.Sprintf("%s/cpu%d/cpufreq/scaling_governor", cpuDir, cpu)
+			err = ioutil.WriteFile(cpuScalingGovernorFile, []byte(governorOrig), 0o644)
+			if err != nil {
+				return err
+			}
+
+			// remove the saved governor
+			err = os.Remove(cpuScalingGovernorFileOrig)
+			if err != nil {
+				return err
```

Comment:
I think we need to ignore `ENOENT` (aka `os.ErrNotExist`) here.

Reply:
I don't think we need to do that - we previously ensure the file exists when we read it above (see line 562). I think it is OK to treat the unexpected disappearance of the file as an error.